### ISSUE TITLE
RFC: `lake exe` should provide suggestions

### ISSUE THREAD

**Issue by semorrison** [leanprover/lean4#2708]
Currently `lake exe` errors with `error: missing executable name`.

Can it list the available `exe`s defined in the current and upstream projects? This would help discoverability.

**Comment by tydeu** [comment]
In this case, should it still print error / return exit code 1 or should it just produce the listing and exit normally?

**Comment by semorrison** [comment]
I think exit code 1 is appropriate, because only humans should ever be doing it: if a script called `lake exe` I'd still want it to fail.

(If there were `lake describe exe` or something, then *that* should exit normally.)

**Comment by tydeu** [comment]
@semorrison 

> (If there were `lake describe exe` or something, then _that_ should exit normally.)

There is `lake help exe`, but loading the workspace as part of help seems like undesirable overhead,

**Comment by sullyj3** [comment]
Is there any reason `lean exe` couldn't just execute the `default_target`?

**Comment by tydeu** [comment]
@sullyj3 It could, but there are some things to consider:

1. Unlike `cargo run`, arguments are passed without a `--` separator, so a `lake exe` with a named exe would have to take no arguments.
2. A package can have multiple executables and multiple default targets, and the default targets do not have to be executables. Thus, it could be quite unclear what executable would run.

**Comment by digama0** [comment]
> Unlike `cargo run`, arguments are passed without a `--` separator, so a `lake exe` with a named exe would have to take no arguments.

In many/most CLIs the `--` separator is only for disambiguation purposes and can be omitted. So you could still allow `lake exe <args>`, provided that the `<args>` are not interpretable as lake arguments or an exe name. And `--` would be available when this results in an incorrect interpretation, i.e. `lake exe -- <args>` unambiguously calls the default target with `<args>`.

Re: 2, you could have another attribute or workspace setting for defining the default executable in ambiguous cases.

**Comment by sullyj3** [comment]
This is how it works in cargo: (from the man page)
```
Target Selection
       When no target selection options are given, cargo run will run the binary target. If there are
       multiple binary targets, you must pass a target flag to choose one. Or, the default-run field
       may be specified in the [package] section of Cargo.toml to choose the name of the binary to run
       by default.

       --bin name
           Run the specified binary.
```
I think this behaviour makes a lot of sense; executing the sole exe of your project if it is the sole exe, and only requiring you to specify in case of ambiguity

**Comment by digama0** [comment]
This does make the very common invocation `lake exe cache get` significantly longer though: `lake exe --bin cache -- get`

**Comment by sullyj3** [comment]
> This does make the very common invocation lake exe cache get significantly longer though: lake exe --bin cache -- get.

I'm not necessarily suggesting the `--bin` flag, just the selection in case of no exe specified.

**Comment by sullyj3** [comment]
how about:
- `lake exe cache get` continues to work as is
- `lake exe someArg` intending to execute the sole executable passing `someArg` should fail with a message along the lines of
> error: unknown executable `someArg`
 If you meant to pass `someArg` as an argument to `my-sole-exe`, run `lake exe -- someArg`

**Comment by nomeata** [comment]
Is that complexity really helpful, compared to the straight forward status quo? The current design is more uniform, works independent whether you are in the project of the executable or a downstream project, and you can add more executables to your project without breaking existing scripts/muscle memory.

**Comment by digama0** [comment]
If it is possible to select the default executable, then it is possible to add executables without breaking existing scripts / muscle memory. If there is only one executable though I agree there is no need to set a default.

**Comment by sullyj3** [comment]
I think there should be a strong presumption towards making the overwhelmingly common case convenient. The overwhelmingly common case is running the program you're writing.

Typing out the name of my exe is annoying, and the fact that it's trivial to infer adds insult to injury. In the scheme of things this is a minor issue. However, in my opinion small inconveniences add up, and they should be eliminated wherever they arise in order to make the overall experience attractive. I think this is the sort of thing that turns people off of established languages. Lean is young, so we put up with mild annoyances temporarily. But I think the aim should be to make UX seamless if the goal is to attract users.

**Comment by joneugster** [comment]
Regarding the initial request of this issue, would it be possible to add something like `lake exes` (or whatever the best naming is) which lists all executable names from the current workspace? Im thinking of something like the following implementation (copied from `lake scripts`):


```lean
protected def exes : CliM PUnit := do
  processOptions lakeOption
  let config ← mkLoadConfig (← getThe LakeOptions)
  noArgsRem do
    let ws ← loadWorkspace config
    for pkg in ws.packages do
      for exe in pkg.leanExes do
        IO.println exe.config.name

```

if I understand things correctly this should be quite simple to implement and would be a great feature to have :)

**Comment by tydeu** [comment]
@joneugster Yes! A minor note: one would probably want to print the executable's full name (i.e., `pkg/exe`) rather than just the executable name (which is what `exe.config.name` is, unlike `script.name`).

**Comment by SephVelut** [comment]
With Rust I am able to use an editor script that I can use to fire off `cargo run` quickly and repeatedly with a keybind to frequently take a peek at a program's output including print statements. This works because often an editor's command line current directory is set to the workspace directory. With lake I cannot do this, because each project requires it's own unique `lake exe [name]`. 

Executing a program you are writing at any time with the least friction should be a priority over all other build ergonomics. Perhaps we could introduce a new command `lake run`.
